### DISCUSSION TITLE
A Rhino wishlist

### DISCUSSION THREAD

**Discussion by bjhess** [KonnorRogers/rhino-editor#208] [Ideas]
We're using Rhino for the [Pika](https://pika.page/) editor. We're thankful that Rhino exists–it has helped us create an editing experience that is very well-loved by our customers. Yet there's still lots we'd like to improve in the Pika editing experience.

Over [here](https://github.com/KonnorRogers/rhino-editor/issues/204#issuecomment-2418044314) @KonnorRogers asked:

> If you have any comments, concerns, any particular features you're looking for, gripes, etc.

I figured a discussion might be a better place to post these things. So here we are!

I am going to talk through the top desires and requests we have for the editor. I realize some/many of these things might be best served as Tiptap extensions. Bullet list time!

* Bubble menu* (we'll get updated to the latest gem soon!)
* Alt Text interface for photos. While most people don't have a clue about Alt Text, those that do are understandably passionate. We currently have code to push a caption into the alt text, but that's not quite right.
* Footnotes, which I've tried to bring in via [this extension](https://github.com/buttondown/tiptap-footnotes). I've not been able to figure it out, but it's more likely a me problem than a Rhino challenge.
* Image resize and alignment. [This](https://www.npmjs.com/package/tiptap-extension-resize-image?activeTab=readme) looks like the most promising extension, but with ActionText's differing markup for images we'd have to recreate it for Rhino, I think.
* Click image to zoom.
* Enhanced Markdown support. We really like what exists in Tiptap for Markdown support, but would like to also see the following supported:
  * Linking
  * Image linking
  * Footnotes
  * Tables, maybe?
* Support the same Markdown when pasting. We actually have some of this in place via some Prosemirror hacking, but this was built by not-me and I can't say that I really understand it. 😬 
* Links in image captions
* Link full images. Not sure how this would interact with a zoom interface.
* An embed interface. Should be an option you have to turn on. Allow embedding of iframes from YouTube, Bandcamp, Spotify, etc. We've made use of the [YouTube Tiptap extension](https://tiptap.dev/docs/editor/extensions/nodes/youtube), which is nice enough. Building similar "paste a URL for embed" support across many services might be a nicer interface, though it would be more work. Possibly building it for 5 example services and making it clear how others could contribute updates for their favorite service would work?
* Automatic anchor tag IDs for headings.
* Details elements (currently a [Tiptap Pro extension](https://tiptap.dev/docs/editor/extensions/nodes/details))
* Tables* seems like it could be a pretty big challenge
* Divider/HR button in toolbar
* Add h1 - h6* is something we've enabled via Markdown using the [Tiptap extension](https://tiptap.dev/docs/editor/extensions/nodes/heading), though we haven't implemented the toolbar options to create h2-h6. Would be nice to have it built in.
* Some bugs here and there, but I'm not prepared at the moment to confirm they are bugs in Rhino vs Tiptap vs our own code. 😆 

\* Also in the [1.0 checklist](https://github.com/KonnorRogers/rhino-editor/issues/197)

I would be happy to help make some of these things a reality in Rhino (or Tiptap). Where we have consensus it seems better to make improvements to the library rather than hacking them in just in our app. However, I'm not very comfortable in JS-land, so any help I can provide will be very slow to arrive. 😆

**Comment by KonnorRogers** [comment]
Please do not apologize! This is immensely helpful!

Let me do my best to address these issues and some of the challenges and ideas I have:

> Alt Text interface for photos. While most people don't have a clue about Alt Text, those that do are understandably passionate. We currently have code to push a caption into the alt text, but that's not quite right.

This is definitely doable. I actually had an idea for this by adding an `<input>` when you have the image "focused" to be able to add plaintext alt text. It is a weird problem because captions aren't always the same as the alt text describing the image.

> Footnotes, which I've tried to bring in via [this extension](https://github.com/buttondown/tiptap-footnotes). I've not been able to figure it out, but it's more likely a me problem than a Rhino challenge.

I'd love to come up with first party support for this. I hadn't even thought of footnotes.

> Image resize and alignment. [This](https://www.npmjs.com/package/tiptap-extension-resize-image?activeTab=readme) looks like the most promising extension, but with ActionText's differing markup for images we'd have to recreate it for Rhino, I think.

This has been on my wishlist for a while. Would also love to be able to "float" images as well. It is quite challenging to get right and I have attempted this once before, although not for very long. I _think_ its doable. I just don't know if I will need to patch ActionText, or if I can come up with a clever workaround.

> Links in image captions

Should've filed an issue for this one! A big thing I wanted to differentiate from Trix is supporting rich text in captions. I know the issue, I'll work on a bug fix.

Issue filed here:

> Enhanced Markdown support. We really like what exists in Tiptap for Markdown support, but would like to also see the following supported:
>
>    Linking
>    Image linking
>    Footnotes
>    Tables, maybe?

Making sure we're not getting terminology crossed here. When you say "markdown", do you mean "markdown inputs" that get translated to HTML by TipTap? Or are you using https://github.com/aguingand/tiptap-markdown ?


> Support the same Markdown when pasting. We actually have some of this in place via some Prosemirror hacking, but this was built by not-me and I can't say that I really understand it. 😬

I'd love to see what you have so I can better understand the request! Even just a video would work!

> Link full images. Not sure how this would interact with a zoom interface.

Not sure I follow what a "zoom interface" is. But linking to external images should be doable. Have to double check ActionText and see what has to happen to support it. And will probably want this to be something people "opt in" to and perhaps have an "allowlist" of domains. Security is always a concern with this stuff.

> An embed interface. Should be an option you have to turn on. Allow embedding of iframes from YouTube, Bandcamp, Spotify, etc. We've made use of the [YouTube Tiptap extension](https://tiptap.dev/docs/editor/extensions/nodes/youtube), which is nice enough. Building similar "paste a URL for embed" support across many services might be a nicer interface, though it would be more work. Possibly building it for 5 example services and making it clear how others could contribute updates for their favorite service would work?

I've talked with @afomera a little about this. [IFramely](https://iframely.com/) seems pretty solid. I'd love to think through what an integration would look like for this that could support a number of services in a single API. Embeds are a very core part of rich text editing.

> Automatic anchor tag IDs for headings.

Oh this is a good one!

> Details elements (currently a [Tiptap Pro extension](https://tiptap.dev/docs/editor/extensions/nodes/details))

Hmmm...I'm curious how they're doing it. It may be doable, but definitely not part of the "core" editor, but maybe available as an extension.

> Tables* seems like it could be a pretty big challenge

Tables. Why'd it have to be tables. There's an issue here of how a team built Tables into their editor. So its definitely doable, and not by using "Custom Attachments" like you would in Trix.

https://github.com/KonnorRogers/rhino-editor/issues/106

Hardest part about tables is getting the UX experience "right". 

> Divider/HR button in toolbar
> Add h1 - h6* is something we've enabled via Markdown using the [Tiptap extension](https://tiptap.dev/docs/editor/extensions/nodes/heading), though we haven't implemented the toolbar options to create h2-h6. Would be nice to have it built in.


So, I have this core philosophy about the editor that the initial editor should not require any changes to ActionText out of the box, and as such introducing these tags to the "base" / "core" experience means "configuring" ActionText, maybe this is misguided and the wrong thing to do, especially as I start making a RhinoEditor gem for form helpers and such, but I do want to keep the core base experience lean, and then make the experience of adding extensions "seamless". Also, you're free to tell me I'm wrong here 😆 I'm just one person and definitely fallible.


> Some bugs here and there, but I'm not prepared at the moment to confirm they are bugs in Rhino vs Tiptap vs our own code. 😆

Bugs?! In a `contenteditable` WYSIWYG?! No! Not possible!

Seriously, always happy to help when you can get something together in an issue!

**Comment by bjhess** [comment]
I've been trying to set up the repo, but am having some challenges. Is the Discord the best place to get some help working through them? On to the response…

> [Alt text] is a weird problem because captions aren't always the same as the alt text describing the image.

For sure. (Side note: Alt text is one of the places where I'm excited for AI to help me out. When it hits an image description it's 👩‍🍳💋.)

>> Links in image captions
>
> I know the issue, I'll work on a bug fix.

:heart:

> When you say "markdown", do you mean "markdown inputs" that get translated to HTML by TipTap? Or are you using https://github.com/aguingand/tiptap-markdown ?

Hmm, this gets into code I didn't write, but:
* In the editing experience, I'm just talking "markdown inputs" that get translated to HTML by Tiptap.
* We have our own "PastedMarkdown" extension that handles a user pasting in Markdown. It borrows heavily from [clipboard.js in tiptap-markdown](https://github.com/aguingand/tiptap-markdown/blob/main/src/extensions/tiptap/clipboard.js), but uses [markdown-it](https://www.npmjs.com/package/markdown-it) for processing.

I think the first point is what pertains to this discussion.

>> Support the same Markdown when pasting.
>
> I'd love to see what you have so I can better understand the request! Even just a video would work!

Here's the "PastedMarkdown" extension that we add to rhino at initialization time: https://gist.github.com/bjhess/57fd02e853c062b116e5923cb5952c7e (I did not write this code and thus deserve no credit for it. :)

>> Link full images. Not sure how this would interact with a zoom interface.
> 
> Not sure I follow what a "zoom interface" is. But linking to external images should be doable. Have to double check ActionText and see what has to happen to support it. And will probably want this to be something people "opt in" to and perhaps have an "allowlist" of domains. Security is always a concern with this stuff.

Yeah, I think I may not have been clear. Some things we'd like:
* The ability to click an image to get a zoomed view of it. As an example, we really like what we've done with the [album covers in Album Whale](https://albumwhale.com/albums/11341). Click the cover to get a closer look at it.
* The ability to not just upload an image, but also to render an image from another domain. So getting an `<img>` tag on screen. (And also being able to do the markdown version of that with `![Alt text](URL)`.)
* The ability to make an image (uploaded or an image from another domain) link to a URL. This is less important to me, but I know some people would like it. Arguable whether it should be implemented as it's kind of confusing IMO.

> I've talked with @afomera a little about this. [IFramely](https://iframely.com/) seems pretty solid. I'd love to think through what an integration would look like for this that could support a number of services in a single API. Embeds are a very core part of rich text editing.

That sounds pretty sweet!

> Tables. Why'd it have to be tables.

That does look pretty complex and impressive! It isn't too highly demanded of a feature for us yet.

> I have this core philosophy about the editor that the initial editor should not require any changes to ActionText out of the box…Also, you're free to tell me I'm wrong here 😆 I'm just one person and definitely fallible.

Heh, I'm in no position to tell you you're wrong!

Thank you!

**Reply by KonnorRogers** [reply]
@bjhess yea discord is probably your best bet, i may need to fix up the README for installation / setup, the setup definitely isn't the easiest thing in the world.

**Comment by bjhess** [comment]
Ah, one more detail about bubble menu. Our ideal bubble menu would be one that can handle the link dialog right within the menu. Why? (I'm playing around with the sample editor on the Rhino website.)

- It's a bit odd for the input to be distant from the text the user is working with, especially on mobile. Though it appears the browser focuses and moves to the link entry field, which is good.
- Though the focus doesn't appear to work consistently? I'm clicking the bubble menu on a previously-linked bit of text and it doesn't seem to focus the dialog consistently. Also the URL isn't in the dialog when I get there. Let me know if you don't see the same thing.
- The big feature for a link in a bubble menu, though, is to be able to preview the link (click to open in new tab). "Checking my links" is a pretty common workflow for me when editing. Deleting the link from the bubble menu would also be nice as well. Kind of like the GDoc interface: ![CleanShot 2024-10-25 at 15 53 05@2x](https://github.com/user-attachments/assets/5640868b-a53f-4cc0-859b-08c990143229)

**Reply by KonnorRogers** [reply]
Oh this was already on my todo list to do revamp the UX around linking. Its quite janky and yup I see the same thing around not persisting the current URL, thanks for documenting it!

**Comment by bjhess** [comment]
Here's an interesting one: how might you support soft returns on mobile when you don't have access to shift+return?

I feel like this has been talked about, but I can't find it. When clicking the link toolbar button (in the toolbar or bubble menu), the highlighted text is not preserved such that when entering the link into the input you've lost track of what text you are actually linking.

**Reply by KonnorRogers** [reply]
I'm not sure I fully understand "soft return" in this case? You have an example of what you would expect it to do?

Are you looking to insert a `<br>` or `<wbr>` or similar on a "soft return"?

and link selection maintaining the selection background has been discussed, I got 0.13.x out to fix some other lingering bugs, but maintaining link selection is the next thing I'm tackling. Requires mucking around with decorators, but should be possible.

**Reply by KonnorRogers** [reply]
@bjhess here's a quick prototype I threw together, still gotta polish it.

https://github.com/user-attachments/assets/185bd3be-01c8-4c90-9837-f3f51212f8b4

**Comment by walterdavis** [comment]
This is often used when typesetting: a linefeed without the New Paragraph semantics. 

A paragraph should have some defined vertical space separating it from other paragraphs in the same parent element. But sometimes you really want a new line without that -- in a poem or an address block, for two good examples. In Markdown, you add two spaces to the end of the line and press return to get this. The output is a <br> rather than </p><p>.

Walter

**Reply by bjhess** [reply]
Yes, poetry is exactly the use case! Works great on a computer, but we've had a customer wonder how they can do the same thing on mobile. It's certainly an edge case, but would be cool to offer an option.

**Comment by bjhess** [comment]
One thing we might attack in 2025 is adding dropdown menus to our Rhino toolbar. Basically we don't want horizontal scrolling while browsing on PC, which means we've reached our max toolbar items. So we need to make some dropdowns.

Curious if that was something you had planned for Rhino as we'd rather not invent our own technique that conflicts with your plans.

**Reply by KonnorRogers** [reply]
Undecided so far as to anything baked in.

I built out a POC of "dropdowns" here: https://role-components.vercel.app/components/menu/

Ideally, people can bring their own dropdowns.

Right now many people have just been using the Shoelace dropdowns https://shoelace.style/components/dropdown/ 

As for dropdowns in the toolbar, probably not something I plan for adding for the current toolbar, except for maybe adding a "Text" dropdown that lets you select "Paragraph, h1, h2, h3", but thats more of a "select" than a "menu".

**Reply by bjhess** [reply]
Thank you for your thoughts!

**Comment by bjhess** [comment]
@KonnorRogers, I'm curious about this particular thing:

>> Footnotes, which I've tried to bring in via [this extension](https://github.com/buttondown/tiptap-footnotes). I've not been able to figure it out, but it's more likely a me problem than a Rhino challenge.
> 
> I'd love to come up with first party support for this. I hadn't even thought of footnotes.

We've actually gotten that extension working in Pika, but there are a lot of flakey bits that have us a little concerned. Weird tabbing behavior, some various breakages, etc.

If you have footnotes on your "somewhat soon" docket, I think we'd prefer to hold off. But if it's more "someday/maybe", we'd probably consider bringing in that Tiptap extension and maybe looking for a way to help shore it up.

**Reply by KonnorRogers** [reply]
its not really super high priority to me, and would probably be a plugin and not part of the main editor. its mostly just a "someday / maybe"

**Comment by bjhess** [comment]
I also wanted to mention:

>> Alt Text interface for photos. While most people don't have a clue about Alt Text, those that do are understandably passionate. We currently have code to push a caption into the alt text, but that's not quite right.
>
> This is definitely doable. I actually had an idea for this by adding an <input> when you have the image "focused" to be able to add plaintext alt text. It is a weird problem because captions aren't always the same as the alt text describing the image.

We're getting a fair bit of heat about our lack of Alt Text. I will probably start digging into a Rhino fork early in the new year to see if I can figure something out to propose. (Though I don't know the first thing about how to do it/bend Rhino to my will, so we'll see!)

I read and considered this [blog post](https://angelika.me/2023/02/26/how-to-add-editing-image-alt-text-tiptap/) as a way to write an extension, but it just seems that given it's handling images in particular (yay, Trix!) it makes the most sense to go directly into Rhino.

**Reply by KonnorRogers** [reply]
TBH, I can probably prioritize this and add it. Give me maybe a week and I should be able to get a POC to play with.

**Reply by KonnorRogers** [reply]
@bjhess if you have any particular designs or UX ideas, happy to use them. The buttons on the blog post are an interesting way i hadn't thought of.

**Reply by bjhess** [reply]
That's awesome, @KonnorRogers! Let me know if you need any help.

Konnor, I think the blog post interface is pretty close. But it might be a little louder than necessary. For some reason I have a vague perception that Mastodon has thought a lot about this. And I notice Bluesky's implementation is similar. What they each have is an alt button over top of the image with an associated indicator of whether or not the image has alt text. I think that's pretty good.

I don't think the button needs to display all of the alt text within itself. First, it's loud. Second, it gets confusing as to whether or not it's a button (thus the need for an "edit" link in that blog post).

It would be kind of nice to optionally have the button only appear on hover, but I think that in itself is likely inaccessible. So probably not a great idea.

Another thing for accessibility is probably to be mindful about how the Alt button is navigated to via tabbing.

Here's a GIF of the Bluesky interaction in action:

![CleanShot 2024-12-18 at 16 51 12](https://github.com/user-attachments/assets/b16908e1-cce8-4f04-a434-516f0a9a8d0d)

**Reply by bjhess** [reply]
Just checking in, @KonnorRogers, to see if this is still in your near-future plans?

**Reply by KonnorRogers** [reply]
Yes, got side tracked by holidays, sorry about that. I got a fairly minimal prototype, I have to add docs, and its not super extensible, and because TipTap / ProseMirror override `mousedown`, I had to do some tweaks to the `<dialog>`, but hopefully will have something soon.

**Reply by bjhess** [reply]
No apologies necessary! Just checking in. Curious to see what you've come up with. Thank you!

**Reply by KonnorRogers** [reply]
Just a little update, I managed to get the TipTap / ProseMirror side 99% working (few issues with clicks not always registering, but workable)

Today I'm working on spelunking through ActionText and how to modify the TrixAttachmen to add an extra attribute for `alt` or `altText` or `alt_text`, any of the above.

Once I figure out how to add additional attributes to the Image attachments, I can store the alt text in the Rails backend. But digging through ActionText attachables and trying to find how to add additional attributes is proving to be a little time consuming, and also means the default editor won't ship with alt text (At least until I make a proper Rhino gem that can ship an initializer to patch it)

soooo...thats where I'm at.

https://github.com/rails/rails/blob/86312f5dc05b96d9c1c71ef03d257c155622f00e/actiontext/lib/action_text/trix_attachment.rb#L10-L18

https://github.com/rails/rails/blob/main/actiontext/lib/action_text/attachable.rb

https://github.com/rails/rails/blob/main/actiontext/lib/action_text/attachables/remote_image.rb

I'm not quite sure which spot is the right one to tackle, so if you have any ideas of how I can add a key / value to `data-trix-attributes` and have the attributes get picked up, im all ears, if not, I'm going to continue hacking away until I figure out what works.

EDIT: I got this working. Will share more when I'm back at a computer.

**Reply by KonnorRogers** [reply]
Easier than expected:

```rb
# config/initializers/actiontext_patch.rb
attributes = ActionText::TrixAttachment::ATTRIBUTES + ["alt"]
ActionText::TrixAttachment.const_set("ATTRIBUTES", attributes)

attributes = ActionText::Attachment::ATTRIBUTES + ["alt"]
ActionText::Attachment.const_set("ATTRIBUTES", attributes)
```

```diff
<!-- app/views/active_storage/blobs/_blob.html.erb -->
<figure class="attachment attachment--<%= blob.representable? ? "preview" : "file" %> attachment--<%= blob.filename.extension %>">
  <% if blob.representable? %>
-   <%= image_tag(blob.representation(resize_to_limit: local_assigns[:in_gallery] ? [ 800, 600 ] : [ 1024, 768 ])) %>
+   <% blob_attributes = blob.full_attributes %>
+   <%= image_tag(blob.representation(resize_to_limit: local_assigns[:in_gallery] ? [ 800, 600 ] : [ 1024, 768 ]), alt: blob_attributes["alt"]) %>
  <% end %>

  <% caption = blob.try(:caption) %>
  <figcaption class="attachment__caption <%= caption ? "attachment__caption--edited" : "" %>">
    <% if caption %>
      <%= caption.html_safe %>
    <% else %>
      <span class="attachment__name"><%= blob.filename %></span>
      <span class="attachment__size"><%= number_to_human_size blob.byte_size %></span>
    <% end %>
  </figcaption>
</figure>
```

Will file upstream with Rails if we can get proper hooks instead of MonkeyPatching.

**Reply by KonnorRogers** [reply]
Quite a bit to cleanup and document, but little demo that it is possible:

https://bsky.app/profile/konnorrogers.com/post/3lfe753wzsc2i

(file is 32mb so cant upload to GitHub)

**Reply by bjhess** [reply]
Thanks, @KonnorRogers! I agree that it'd be great to not have to monkey patch this, but I wonder what the odds are that that is going to be possible.

Let me know if you need anything!

**Comment by bjhess** [comment]
In my ongoing efforts to "influence" 😈  the next thing you work on in Rhino, I'm curious what you think of a zoom interface for clicking images? For example how you can [click a cover on Album Whale](https://albumwhale.com/albums/19187) to see a bigger version of the image. If this isn't in your vision for Rhino, we'll start thinking on another way.

**Reply by KonnorRogers** [reply]
Do you mean zoom in the editor interface? Or zoom in the final output?

Zoom in the editor is hard because of the number of existing interactions, and seems largely...unnecessary??

Zoom in the final output would be out of scope of the editor, but I could put together docs and a code snippet of how you could do it.

**Reply by bjhess** [reply]
Sorry, yes, zoom in the final output. If you have thoughts I’d appreciate seeing them, but totally up to you! Thanks!

**Reply by KonnorRogers** [reply]
Here's a video of it:


https://github.com/user-attachments/assets/c7f0cc63-fc28-4c45-b500-aa830b283690

And the JavaScript:

```js
function handleLightBox (e) {
  const target = e.target

  // Dont show light boxes in the editor
  if (target.closest("rhino-editor")) { return }

  // Dont show light boxes if not inside a trix rendered output
  if (!target.closest(".trix-content")) { return }

  // action-text-attachment only appears in final output, not in editor, so this is just an extra check.
  const attachment = target.closest("action-text-attachment")
  const image = target.closest("img")

  // Check to make sure someone actually clicked the image.
  if (!image || !attachment) {
    return
  }

  const dialog = document.createElement("dialog")
  dialog.classList.add("lightbox")
  const clonedImage = image.cloneNode(true)
  dialog.append(clonedImage)
  clonedImage.style.setProperty("--aspect-ratio", `${image.naturalWidth} / ${image.naturalHeight}`)
  document.body.append(dialog)
  dialog.showModal()

  dialog.addEventListener("close", () => {
    dialog.remove()
  })
}

document.addEventListener("click", handleLightBox)
```

and finally the CSS:

```css
<style>
  .trix-content .attachment--preview img:hover {
    cursor: zoom-in;
  }

  .lightbox {
    height: 75vh;
    max-height: min(75vh, 600px);
    width: 75vw;
    max-width: min(75vw, 800px);
    display: grid;
    place-content: center;
  }

  .lightbox img {
    display: block;
    height: auto;
    margin: auto;
    width: 100%;
    max-width: 100%;
    max-height: 100%;
    aspect-ratio: var(--aspect-ratio, 16 / 9);
  }
</style>
```

And here's a branch where I did a little test run:

https://github.com/KonnorRogers/rhino-editor/tree/konnorrogers/test-out-lightbox-implementation

I think that should square you away. Let me know if you have any questions!

**Reply by bjhess** [reply]
Thanks for this, BTW. Ended up using [zoom-vanilla](https://github.com/spinningarrow/zoom-vanilla.js).

Decided to silly-like update all the `trix-content img` tags with JS. Could have used `active_storage/blobs/_blot.html.erb` to add the required `data-action` attribute to `img`, but this allows us to control which `.trix-content` areas zooming applies to with an extra `.zoomable` class.

```js
import { Controller } from "@hotwired/stimulus"
import zoom from "zoom-vanilla.js"

export default class extends Controller {
  connect() {
    const images =
      document.querySelectorAll('.zoomable .trix-content img')

    images.forEach(img => {
      img.setAttribute('data-action', 'zoom')
    })
  }
}
```

```scss
// Copied from zoom-vanilla:
// https://github.com/spinningarrow/zoom-vanilla.js/blob/master/css/zoom.css

img[data-action="zoom"] {
  cursor: zoom-in;
}

img[data-action="zoom-out"] {
  cursor: zoom-out;
}

.zoom-img,
.zoom-img-wrap {
  position: relative;
  transition: all 300ms;
  z-index: 666;
}

.zoom-overlay {
  background: var(--color-bg);
  bottom: 0;
  left: 0;
  opacity: 0;
  pointer-events: none;
  position: fixed;
  right: 0;
  top: 0;
  transition: opacity 300ms;
  z-index: 665;
}

.zoom-overlay-open .zoom-overlay {
  opacity: 0.9;
}

.zoom-overlay-open,
.zoom-overlay-transitioning {
  cursor: default;
}
```

**Comment by bjhess** [comment]
> Automatic anchor tag IDs for headings.

Here's one approach I'm considering deploying to Pika. It borrows from the ideas in syntax highlighting by just writing the IDs to the doc before it is submitted to the DB:

https://gist.github.com/bjhess/3ec812e096414dee0c2ecfc9f285fd1d

I think long-term it'd probably be nicer to extend Tiptap's Heading extension so the IDs are added real-time as you type in the editor. Two reasons I can see for this:

- Eventually, provide a UI where you can copy the link to that heading for pasting elsewhere in the document
- Perhaps as a hook for a TOC (table of contents) extension later that would allow people to have a dynamic TOC at the top of the page that links to all of the headers within the page

I'm not seeing a big conflict between this doc-rewrite concept and an eventual in-editor solution. Seems you could just replace the former with the latter.

**Reply by KonnorRogers** [reply]
Good news is that its definitely possible:

https://tiptap.dev/docs/editor/extensions/functionality/table-of-contents

bad news is that its a private extension. _But_ I think an "appendTransaction" could very easily handle this.

**Reply by bjhess** [reply]
Yeah, we'll need to multiply our user count a bit before I jump into the private extensions. Maybe someday! 🤞

**Reply by bjhess** [reply]
I have this basically built using the general concept above.

Now to consider how to allow `#anchor-links` in the link dialog box…

**Reply by KonnorRogers** [reply]
not sure I follow...do you have a rough visual I could see?

**Reply by bjhess** [reply]
Given there are anchor ids on the page, when linking to them within the same page (say if you're making a table of contents), it would be best to do it like [this](https://stackoverflow.com/a/12096697) (e.g. `<a href="#anchor-link">`). But that's not a valid link that can be entered in the URL field:

![CleanShot 2025-05-02 at 15 06 59@2x](https://github.com/user-attachments/assets/8883a162-d4ca-4c16-a01a-5550d55aed78)

Something I'll need to think about as a future enhancement.

**Comment by bjhess** [comment]
>> Image resize and alignment. [This](https://www.npmjs.com/package/tiptap-extension-resize-image?activeTab=readme) looks like the most promising extension, but with ActionText's differing markup for images we'd have to recreate it for Rhino, I think.
> 
> This has been on my wishlist for a while. Would also love to be able to "float" images as well. It is quite challenging to get right and I have attempted this once before, although not for very long. I think its doable. I just don't know if I will need to patch ActionText, or if I can come up with a clever workaround.

Just saw this demo and thought of the resize discussion:

https://github.com/HMarzban/extension-hypermultimedia?tab=readme-ov-file#demo-time-

**Reply by bjhess** [reply]
@KonnorRogers, I'm curious if you are considering diving into the image resizing feature in the coming months? It continues to be a high-demand request for our Pika customers, so I just wanted to check in. Thanks!

**Reply by KonnorRogers** [reply]
@bjhess sorry for leaving this for so long. I'll be honest, I haven't even thought about it. 

With the launching of the Web Awesome app and all my obligations there, and given the current state of the world and people close to me who are affected by the current political atmosphere in the US, I've been significantly burnt out on personal projects and taken a step back for a little bit.

I'll probably come back to working on Rhino in the coming months, but always happy to help guide what an image resizer plugin could look like.

**Reply by bjhess** [reply]
No worries, @KonnorRogers. Take care of you and yours!

**Comment by bjhess** [comment]
An FYI:

>> Details elements (currently a [Tiptap Pro extension](https://tiptap.dev/docs/editor/extensions/nodes/details))
>
> Hmmm...I'm curious how they're doing it. It may be doable, but definitely not part of the "core" editor, but maybe available as an extension.

Near as I can tell Details is no longer a pro extension. Several others have been opened up as well. RE: https://tiptap.dev/blog/release-notes/were-open-sourcing-more-of-tiptap
